File: 1475-final-prices-with-a-special-discount-in-a-shop/1475-final-prices-with-a-special-discount-in-a-shop.py
class Solution:
    def finalPrices(self, prices: list[int]) -> list[int]:
        n = len(prices)
        stack = []
        for i in range(n - 1, -1, -1):
            while stack and stack[-1] > prices[i]:
                stack.pop()
            tmp = prices[i]
            if stack:
                prices[i] -= stack[-1]
            stack.append(tmp)
        return prices

File: 1475-final-prices-with-a-special-discount-in-a-shop/1475-final-prices-with-a-special-discount-in-a-shop.js
var finalPrices = function(prices) {
    const n = prices.length;
    const stack = [];
    for (let i = n - 1; i >= 0; i--) {
        while (stack.length && stack[stack.length - 1] > prices[i]) {
            stack.pop();
        }
        const tmp = prices[i];
        if (stack.length) {
            prices[i] -= stack[stack.length - 1];
        }
        stack.push(tmp);
    }
    return prices;
};
